feugy/tabulous
# fix(web): poor target detection and gravity performance and accuracy.

### What's in there?

While trying out the last private game, Tabulous appeared slow when dragging 5+ cubes or cards.
A quick investigation showed two bottlenecks:
- `Scene.getMeshById()`, which iterates over all the scene meshes to find one
- `Scene.pickWithRay()` when looking for meshes below the moved one(s)

Besides improving these functions, PR fixes many smaller issues:

- fix(web): gravity only considers bounding box edges and center (long vertical meshes are not picked)
- fix(web): no re-ordering animation for stacks of non-flippable/rotable meshes
- fix(web): re-ordering animation does not look nice on wide meshes
- fix(web): when stacked, only the last mesh is actionable
- refactor(web): improves `isAbove()` and `gravity()` performances
- refactor(web): optimizes Babylon's `getMeshById()`
- chore(web): fixes Codacy minor violations
- docs: add readme badges

### How to test?

On our latest private game, try stacking hexagonal and cylindric tokens, as well as cards, and check various edge cases.
Shuffle various stacks of cards and tokens.
Move, open the action menu, and interact with tall stacks of tokens, and a flat camera angle.
Move large selection of meshes, and drop them.

### Notes to reviewers

`Scene.getMeshById()` was already identified as a big bottleneck: Babylon's does not use `Map` to reference meshes because they may not have ids when they are created, and these ids may change in a mesh's lifetime.
In our case, it is safe to assume a mesh id will never change and will be known when building it. The only exception is custom meshes, because `Scene.ImportMesh()` creates them with the ids stored in the `.babylon` file.

To mitigate this, our `createCustom()` builder removes and adds again imported meshes to ensure they are referenced with their desired ids. As a side effect, the hand manager must delay it's hook on `onNewMeshAddedObservable` since names are assigned _after_ the event is triggered.

Detecting mesh below with `Scene.pickWithRay()` triggers unnecessary computations and is not really reliable (as we only check corners of one mesh's bounding box). Instead of casting 3D rays for these peculiar points, we now detects shape overlap on the ground, by intersecting 2D geometric shapes, which is much simpler and faster.
The general solution for mesh is to use their bounding box (super reliable for cards, rounded tiles and boxes).
Mesh new `isCylindric` attributes allows using an enclosing circle (within their bounding box, as opposed to their bounding sphere), when relevant (very suitable for 5+ edge prisms and round tokens).
These are no perfect solutions, since triangular prism will use a circle, and custom meshes a bounding box, but much cheaper and reliable.
